lib/ffmprb/filter.rb

```diff
@@ -0,0 +1,234 @@
+        inout "afade=in:d=#{duration}", input, output
+      end
+
+      def afade_out(duration=1, input=nil, output=nil)
+        inout "afade=out:d=#{duration}", input, output
+      end
+
+      # NOTE inputs order matters
+      def amix(inputs, output=nil)
+        inout "amix=#{[*inputs].length}:duration=first", inputs, output
+      end
+
+      def anull(input=nil, output=nil)
+        inout "anull", input, output
+      end
+
+      def anullsink(input=nil)
+        inout "anullsink", input, nil
+      end
+
+      def asplit(inputs=nil, outputs=nil)
+        inout "asplit", inputs, outputs
+      end
+
+      def atrim(st, en, input=nil, output=nil)
+        inout "atrim=#{[st, en].compact.join ':'}, asetpts=PTS-STARTPTS", input, output
+      end
+
+      def black_source(duration, resolution=nil, fps=nil, output=nil)
+        filter = "color=black:d=#{duration}"
+        filter << ":s=#{resolution}"  if resolution
+        filter << ":r=#{fps}"  if fps
+        inout filter, nil, output
+      end
+
+      def fade_out_alpha(duration=1, input=nil, output=nil)
+        inout "fade=out:d=#{duration}:alpha=1", input, output
+      end
+
+      def fps(fps, input=nil, output=nil)
+        inout "fps=fps=#{fps}", input, output
+      end
+
+      def concat_v(inputs, output=nil)
+        inout "concat=#{[*inputs].length}:v=1:a=0", inputs, output
+      end
+
+      def concat_a(inputs, output=nil)
+        inout "concat=#{[*inputs].length}:v=0:a=1", inputs, output
+      end
+
+      def concat_av(inputs, output=nil)
+        inout "concat=#{inputs.length/2}:v=1:a=1", inputs, output
+      end
+
+      def copy(input=nil, output=nil)
+        inout "copy", input, output
+      end
+
+      def crop(crop, input=nil, output=nil)
+        inout "crop=#{crop_exps(crop).join ':'}", input, output
+      end
+
+      def crop_exps(crop)
+        exps = []
+
+        if crop[:left] > 0
+          exps << "x=in_w*#{crop[:left]}"
+        end
+
+        if crop[:top] > 0
+          exps << "y=in_h*#{crop[:top]}"
+        end
+
+        if crop[:right] > 0 && crop[:left]
+          raise Error, "Must specify two of {left, right, width} at most"  if crop[:width]
+          crop[:width] = 1 - crop[:right] - crop[:left]
+        elsif crop[:width] > 0
+          if !crop[:left] && crop[:right] > 0
+            crop[:left] = 1 - crop[:width] - crop[:right]
+            exps << "x=in_w*#{crop[:left]}"
+          end
+        end
+        exps << "w=in_w*#{crop[:width]}"
+
+        if crop[:bottom] > 0 && crop[:top]
+          raise Error, "Must specify two of {top, bottom, height} at most"  if crop[:height]
+          crop[:height] = 1 - crop[:bottom] - crop[:top]
+        elsif crop[:height] > 0
+          if !crop[:top] && crop[:bottom] > 0
+            crop[:top] = 1 - crop[:height] - crop[:bottom]
+            exps << "y=in_h*#{crop[:top]}"
+          end
+        end
+        exps << "h=in_h*#{crop[:height]}"
```

Comment:
Both w and h calculation could be dried out by moving the generic logic to a separate method.  It looks like the same code with different params.